### PR TITLE
New fetchAroundLocation, added stuff to the mainscreen.

### PR DESCRIPTION
**Pull request by Athanaze**
This PR implements a simple feed to MainScreen.
It works purely based on location, so the user does not need to be logged in or to have an account to see it.
The "recommendation algorithm" is straightforward:
Give the closest N events near me (using the current GPS location).
For now, N = 4

To achieve this, a new distanceTo method has been added to the Location data class, this allows the events list to be sorted by distance (useful since we want to take the closest N events).

To get the user location, a LocationViewModel has been implemented. It might be useful to also use it in FindEvent. For simplicity, it's currently declared in the same file as HomeScreen.

### CODE REVIEW

**Comment by Athanaze** [comment]
signal-2024-05-08-115321

**Comment by Athanaze** [comment]
image

File: app/src/main/java/com/monkeyteam/chimpagne/ui/components/EventCard.kt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.style.TextOverflow
+import androidx.compose.ui.unit.dp
+import com.monkeyteam.chimpagne.R
+import com.monkeyteam.chimpagne.model.database.ChimpagneEvent
+import com.monkeyteam.chimpagne.model.utils.timestampToStringWithDateAndTime
+import com.monkeyteam.chimpagne.ui.theme.ChimpagneTypography
+
+@Composable
+fun EventCard(event: ChimpagneEvent, modifier: Modifier = Modifier, onClick: () -> Unit) {
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/components/EventCard.kt:40]: The fact that you decided to create a new component is a good practise 👍

File: app/src/main/java/com/monkeyteam/chimpagne/ui/components/EventCard.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                      )
+                }
+          }
+          Column(
+              modifier =
+                  Modifier.fillMaxWidth()
+                      .padding(top = 8.dp, start = 16.dp, end = 16.dp, bottom = 16.dp)) {
+                Text(
+                    text = event.title,
+                    style = ChimpagneTypography.titleLarge,
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/components/EventCard.kt:97]: We discussed it today, and you already moved to ChimpagneTypography. Fast adaptation is the key to success 😎

File: app/src/main/java/com/monkeyteam/chimpagne/viewmodels/FindEventsViewModel.kt [hunk 3]
```diff
@@ -69,6 +69,30 @@ class FindEventsViewModel(database: Database) : ViewModel() {
     }
   }
 
+  fun fetchAroundLocation(onSuccess: () -> Unit = {}, onFailure: (Exception) -> Unit = {}) {
+    eventManager.getAllEventsByFilterAroundLocation(
+        _uiState.value.selectedLocation!!,
+        9999999999.0,
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/viewmodels/FindEventsViewModel.kt:75]: Next time, use a constant or something like `MaxDouble`

File: app/src/test/java/com/monkeyteam/chimpagne/model/LocationTest.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package com.monkeyteam.chimpagne.model
+
+import com.monkeyteam.chimpagne.model.location.Location
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class LocationTest {
```
Comment (arnaudlvq) [inline app/src/test/java/com/monkeyteam/chimpagne/model/LocationTest.kt:7]: Good code coverage

**Comment by Athanaze** [comment]
https://github.com/Chimpagne/ChimpagneApp/assets/29517376/fea9881b-f52d-405a-8660-af64e05cca6c

**Review by arnaudlvq — APPROVED** [review]
The PR needs to be merged because it is lagging behind. Also, you did implement the core features which make this PR acceptable. 

As you did not use the same approch for getting the GPS enabled as in FindEventScreen, we will have work to do on this home page for the next sprint.

We'll have to check for the screen updates and the UI in the futur as well.

But overall, your code is clean, and your PR has a good code coverage. Keep it up !

File: app/src/main/java/com/monkeyteam/chimpagne/viewmodels/FindEventsViewModel.kt (hunk 3 above)
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/viewmodels/FindEventsViewModel.kt:75]: Magic value. Is it really useful to display such far events ? The fetched events are supposed to be *close* to the user.

File: app/src/main/java/com/monkeyteam/chimpagne/model/location/Location.kt [hunk 5]
```diff
@@ -26,4 +26,29 @@ data class Location(
...
       GeoFireUtils.getGeoHashForLocation(GeoLocation(latitude, longitude)))
+
+  /*
+       It applies the Haversine formula to calculate the great-circle distance between the two locations:
+     - It calculates the square of the half-chord length between the two points (`a`) using the Haversine formula.
+     - It calculates the angular distance in radians (`c`) using the inverse Haversine formula.
+
+  4. Finally, it multiplies the angular distance (`c`) by the Earth's radius (`earthRadius`) to obtain the distance in kilometers.
+       */
+  fun distanceTo(other: Location): Double {
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/model/location/Location.kt:37]: good job :clap:

**Pull request merged**